feat: 우선순위 기반 스케줄링 구현

hyunmuam opened the pull request:
## 🔗 관련 이슈
<!-- 이슈 번호를 연결해주세요 -->
closes #10 

## 📌 작업 내용
### 구현 사항
<!-- 구현내용에 대해서 설명해주세요. -->
<!-- 예시 :
- 이슈 템플릿, PR 템플릿 작성
- READEME 파일 최초 작성
 --> 
- 우선순위 기반 스케줄링 구현: `ready_list`를 우선순위 순 정렬
- 세마포어 우선순위 스케줄링 지원: 최고 우선순위 스레드 먼저 깨움
- 조건 변수 우선순위 스케줄링 지원: `signal` 시 최고 우선순위 스레드 선택
- `sleep_list` 정렬 시 동일 `wake_tick`에 대해 `priority` 고려

### 변경된 파일
<!-- 구현 내용으로 인해 변경된 파일에 대해 설명해주세요 -->
<!-- 예시:
- README.md
  - readme 파일 최초 생성
- issue_template.md
  - 이슈 템플릿 최초 생성
-->
- `thread.c`, `thread.h`
  - `cmp_priority()` 전역 함수로 변경
  - `thread_create()`: 새 스레드 우선순위가 높으면 즉시 yield
  - `thread_unblock()`, `thread_yield()`: ready_list 우선순위 순 정렬
- `synch.c`
  - `sema_down()`: waiters에 `list_push_back`으로 삽입
  - `sema_up()`: `list_min`으로 최고 우선순위 스레드 탐색 및 선점
  - `cond_signal()`: waiters 재정렬 후 최고 우선순위 스레드 깨움
  - `cmp_cond_waiter_priority()` 비교 함수 추가
- `timer.c`
  - sleep_list 정렬 기준에 priority 추가

### 추가 사항
<!-- 트러블 슈팅 등 추가로 설명할 내용이 있으시다면 아래에 작성해주세요. -->
- `sema_up()`에서 `list_min()` 사용 이유: `cmp_priority`가 내림차순(>) 비교이므로 `list_min`이 최고 우선순위 반환
- priority donation 대비 삽입 시 정렬 대신 탐색 시 최고 우선순위 선택 방식 채택

## ✅ 셀프 체크리스트
<!-- 체크리스트를 모두 완료한 후 제출해주세요 -->
<!-- 완료 처리를 표시하려면 [x], 완료되지 않았다면 [ ] -->
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 주석이나 디버깅 코드를 제거했나요?
- [x] 관련 테스트가 모두 통과하나요?
- [x] 커밋 메시지가 컨벤션에 맞나요?

hyunmuam committed:
feat: sleep_list 정렬 기준에 priority 추가

- wake_tick이 동일한 경우 priority를 고려하여 정렬하도록 변경.

hyunmuam committed:
feat: 우선순위 기반 스케줄링을 위한 ready_list 정렬 구현

- thread_create() 시 새 스레드가 현재 스레드보다 우선순위가 높으면 즉시 yield
- thread_unblock()과 thread_yield()에서 list_insert_ordered()를 사용해 ready_list를 우선순위 순으로 유지
- 우선순위 비교 함수(cmp_priority) 추가
- ready_list의 begin이 항상 최고 우선순위 스레드가 되도록 보장
- 이를 통해 next_thread_to_run()이 O(1)로 최고 우선순위 스레드를 선택

hyunmuam committed:
style: synch.c, timer.c에 변경된 포매팅 규칙 적용

hyunmuam committed:
feat: 세마포어 우선순위 기반 대기/해제 구현

- thread.c의 cmp_priority를 전역 함수로 변경
- sema_down(): waiters에 우선순위 순 삽입
- sema_up(): 높은 우선순위 스레드 unblock 시 즉시 선점

hyunmuam committed:
feat: 조건 변수 우선순위 기반 깨우기 구현

- cmp_cond_waiter_priority() 비교 함수 추가
- cond_signal()에서 waiters 재정렬 후 최고 우선순위 스레드 깨움

hyunmuam committed:
refactor: 세마포어 동적 우선순위 변경 대응

- priority donation에서 대기 중 우선순위 변경 시 정렬 깨짐 문제 방지
- sema_down(): 삽입 시 정렬 제거 (list_push_back)
- sema_up(): list_min으로 현재 최고 우선순위 스레드 탐색

hdh853 commented on devices/timer.c:
```diff
@@ -209,9 +209,16 @@ static void real_time_sleep(int64_t num, int32_t denom)
+    if (ta->wakeup_tick != tb->wakeup_tick)
+        return ta->wakeup_tick < tb->wakeup_tick;
+
+    return ta->priority > tb->priority;
```
깨울때 thread_unblock()을 해서 깨우게 되는데, unblock에서 ready_list에 넣을 때 list_push_back이 아니라 list_insert_ordered 로 넣고 있어서 기존처럼 return ta->wakeup_tick < tb->wakeup_tick; 만 해도 괜찮을 것 같은데, 추가하신 이유가 확장성 때문인지 아니면 로직의 자연스러움 때문인지 궁금합니다!

hyunmuam replied:
@hdh853 
구현 순서에 따라 단편적으로 생각하다 보니 이렇게 작성된 것 같습니다.
다시 보니 말씀해 주신 것처럼 중복해서 정렬하고 있었네요.
어차피 `thread_unblock`에서 `ready_list`에 삽입할 때 우선순위 기준으로 정렬해서 넣기 때문에, 깨우는 시점에서는 따로 우선순위를 기준으로 정렬할 필요는 없을 것 같습니다.
말씀해 주신 부분은 제거하는 방향으로 수정해 보겠습니다.

hdh853 commented on threads/synch.c:
```diff
@@ -295,16 +303,17 @@ cond_wait (struct condition *cond, struct lock *lock) {
+
+    if (!list_empty(&cond->waiters)) {
+        list_sort(&cond->waiters, cmp_cond_waiter_priority, NULL);
+        sema_up(&list_entry(list_pop_front(&cond->waiters), struct semaphore_elem, elem)->semaphore);
```
위의 sema_up에선 '넣을때 그냥 뒤어 넣기/뺄 때 list_min으로 빼기'로 하셨는데,
cond_signal에선 호출될때 마다 sort -> list_pop_front()로 하신 이유가 궁금합니다!

hyunmuam replied:
@hdh853
`list_min`을 사용해서 수정하는 방향이 더 적절할 것 같습니다. ㅎㅎ;;

hdh853 reviewed (commented):
이전에 했던 작업들인데 구현 방식이 조금씩 다른게 신기하네요 ㅎㅎ
전반적으로 깔끔하게 로직을 짜주셔서 보기 편한것 같습니다.
고생하셨습니다!

hyunmuam committed:
refactor: 깨울 때 중복 우선순위 정렬 로직 제거

- thread_unblock()에서 이미 ready_list에 우선순위 기준으로 정렬 삽입(list_insert_ordered)하므로, 스레드를 깨우는 시점에 추가로 우선순위 정렬할 필요 없음
- wakeup_tick_less 비교 함수에서 priority 비교 로직 제거

hyunmuam committed:
refactor: cond_signal을 list_min 기반 탐색으로 변경

- sema_up과 일관되게 list_min으로 최고 우선순위 스레드 탐색
- 매번 정렬하는 대신 탐색 시점에 최고 우선순위 선택
- list_sort + list_pop_front → list_min으로 통일

hdh853 reviewed (approved):
고생하셨습니다!
